crates/walrus-service/src/node/events/event_processor.rs

```diff
@@ -1086,6 +1101,36 @@ impl EventProcessor {
             .collect();
         (first_event, relevant_events)
     }
+
+    /// Updates the cached checkpoint sequence number.
+    fn update_checkpoint_cache(&self, sequence_number: u64) {
+        // Use fetch_max to atomically update the value if sequence_number is greater
+        self.latest_checkpoint_seq_cache
+            .fetch_max(sequence_number, Ordering::SeqCst);
+    }
+
+    /// Gets the latest checkpoint sequence number, preferring the cache.
+    pub fn get_latest_checkpoint_sequence_number(&self) -> Option<u64> {
+        // Try to get from cache first.
+        let cached = self.latest_checkpoint_seq_cache.load(Ordering::SeqCst);
+        if cached > 0 {
+            return Some(cached);
+        }
+
+        // If cache is empty, query the store.
+        let seq_num = match self.stores.checkpoint_store.get(&()) {
+            Ok(Some(checkpoint)) => Some(*checkpoint.inner().sequence_number()),
+            _ => None,
+        };
+
+        // Update the cache if we had to read from the store.
+        if let Some(num) = seq_num {
+            self.latest_checkpoint_seq_cache
+                .store(num, Ordering::SeqCst);
+        }
```

Comment:
I am not sure why we are doing the get and update of the cache here. It seems like cache is always kept up to date in the download loop. Do we also need to update `latest_checkpoint_seq_cache` here?